### PR TITLE
Documentation of Reroute topology.

### PR DESCRIPTION
**Pull request by dpoltavets**
Documents for the task of eliminating Cache topology.
#561

Implementation: #1520 (Implementation of the Reroute topology and NetworkTopologyBolt in the Event topology)

### CODE REVIEW

File: docs/design/solutions/reroute-topology/sequence-reroute-topology.puml [hunk 1]
```diff
@@ -0,0 +1,131 @@
...
+        "ET-NetworkTopologyService" -> kilda.topo.reroute : CommandMessage\nRerouteInactiveFlows
+    end
+end
+
+kilda.topo.reroute -> "RT-RerouteBolt" : CommandMessage\n-RerouteAffectedFlows (contains PathNode)\n-RerouteInactiveFlows
+"RT-RerouteBolt" -> "RT-RerouteService" : handleEvents
+
+activate "RT-RerouteService"
+alt RerouteAffectedFlows
+    "RT-RerouteService" -> DAO : dumpFlows
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/reroute-topology/sequence-reroute-topology.puml]: Yes, we can. I fixed it. And now we should send query to receive affected flows from DB (we do not need to filter all dumped flows).

**Comment by dpoltavets** [comment]
@surabujin Yes, the flow topology(CrudBolt) will use DAO too.

**Comment by dpoltavets** [comment]
@surabujin I changed the sequence diagram. Now there are 3 services that work with the DAO layer and one to send reroute commands.

File: docs/design/solutions/reroute-topology/sequence-reroute-topology.puml [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    activate DAO
+    DAO -> "ET-IslService" : islFromDao
+    deactivate DAO
+
+    "ET-IslService" -> DAO : createOrUpdateIsl
+    "ET-IslService" -> DAO : pullPropertiesFromLinkProps
+    "ET-IslService" -> DAO : updateIslBandwidth
+deactivate "ET-IslService"
+
+"ET-NetworkTopologyBolt" -> "ET-Sender" : sendCommandToRerouteInactiveFlows
```
Comment (surabujin) [inline docs/design/solutions/reroute-topology/sequence-reroute-topology.puml]: You must not emit reroute command until topology change are stored in DB.

File: docs/design/solutions/reroute-topology/sequence-reroute-topology.puml [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+
+Floodlight -> kilda.topo.disco : InfoData:\n-SwitchInfoData\n-IslInfoData\n-PortInfoData
+kilda.topo.disco -> "ET-OfeLinkBolt" : InfoData:\n-SwitchInfoData\n-IslInfoData\n-PortInfoData
+"ET-OfeLinkBolt" -> "ET-NetworkTopologyBolt" : InfoData:\n-SwitchInfoData\n-IslInfoData\n-PortInfoData
+activate "ET-NetworkTopologyBolt"
+
+alt SwitchInfoData
+
+alt ADDED
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/reroute-topology/sequence-reroute-topology.puml]: Now we just transfer logic from TE. In the TE module, these actions are divided. Therefore, here these events are handled separately.

File: docs/design/solutions/reroute-topology/reroute-topology.md [hunk 4]
```diff
@@ -0,0 +1,9 @@
+# Reroute topology: eliminate Cache topology
+
+## The problem 
+One of the things that blocks Hub&Spoke is Cach topology. The main task of Cache topology is rerouting the flow. We need to implement this functionality as a new topology, and the part related to processing events (SwitchInfoData, IslInfoData, FlowInfoData) should be implemented in the Event topology.
```
[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/reroute-topology/reroute-topology.md]: Thank you! You are right.

File: docs/design/solutions/reroute-topology/use-case-reroute-topology.puml [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+rectangle "Event topology" {
+    rectangle "Switch operations" {
+        (Create or update\nswitch) as CUS
+        (Activate switch) as AS
+        (Deactivate switch) as DS
+    }
+
+    rectangle "Isl operations" {
+        (Create or update\nISL) as CUI
+        (Set ISL status "moved") as IDF
```
Comment (surabujin) [inline docs/design/solutions/reroute-topology/use-case-reroute-topology.puml]: FL do not produce "ISL moved" event. Also "ISL moved" is not an ISL operation... this is mostly alias for "ISL DOWN" with some specific precondition.

File: docs/design/solutions/reroute-topology/sequence-reroute-topology.puml (hunk 3 above)
[the remark this answers is not in the document]
Comment (sergii-iakovenko) [reply docs/design/solutions/reroute-topology/sequence-reroute-topology.puml]: @dpoltavets Let's perform these changes in a scope of a separate issue.

[the remark this answers is not in the document]
Comment (dpoltavets) [reply docs/design/solutions/reroute-topology/sequence-reroute-topology.puml]: @sergii-iakovenko I created an issue related to this task: #1521

**Pull request merged**